src/array/cpu/spmat_op_impl.cc

```diff
@@ -0,0 +1,619 @@
+
+///////////////////////////// CSRGetRowNNZ /////////////////////////////
+
+template <DLDeviceType XPU, typename IdType>
+int64_t CSRGetRowNNZ(CSRMatrix csr, int64_t row) {
+  CHECK(row >= 0 && row < csr.num_rows) << "Invalid row index: " << row;
+  const IdType* indptr_data = static_cast<IdType*>(csr.indptr->data);
+  return indptr_data[row + 1] - indptr_data[row];
+}
+
+template int64_t CSRGetRowNNZ<kDLCPU, int32_t>(CSRMatrix, int64_t);
+template int64_t CSRGetRowNNZ<kDLCPU, int64_t>(CSRMatrix, int64_t);
+
+template <DLDeviceType XPU, typename IdType>
+NDArray CSRGetRowNNZ(CSRMatrix csr, NDArray rows) {
+  const auto len = rows->shape[0];
+  const IdType* vid_data = static_cast<IdType*>(rows->data);
+  const IdType* indptr_data = static_cast<IdType*>(csr.indptr->data);
+  NDArray rst = NDArray::Empty({len}, rows->dtype, rows->ctx);
+  IdType* rst_data = static_cast<IdType*>(rst->data);
+  for (int64_t i = 0; i < len; ++i) {
+    const auto vid = vid_data[i];
+    rst_data[i] = indptr_data[vid + 1] - indptr_data[vid];
+  }
+  return rst;
+}
+
+template NDArray CSRGetRowNNZ<kDLCPU, int32_t>(CSRMatrix, NDArray);
+template NDArray CSRGetRowNNZ<kDLCPU, int64_t>(CSRMatrix, NDArray);
+
+///////////////////////////// CSRGetRowColumnIndices /////////////////////////////
+
+template <DLDeviceType XPU, typename IdType>
+NDArray CSRGetRowColumnIndices(CSRMatrix csr, int64_t row) {
+  CHECK(row >= 0 && row < csr.num_rows) << "Invalid row index: " << row;
+  const int64_t len = impl::CSRGetRowNNZ<XPU, IdType>(csr, row);
+  const IdType* indptr_data = static_cast<IdType*>(csr.indptr->data);
+  const int64_t offset = indptr_data[row] * sizeof(IdType);
+  return csr.indices.CreateView({len}, csr.indices->dtype, offset);
+}
+
+template NDArray CSRGetRowColumnIndices<kDLCPU, int32_t>(CSRMatrix, int64_t);
+template NDArray CSRGetRowColumnIndices<kDLCPU, int64_t>(CSRMatrix, int64_t);
+
+///////////////////////////// CSRGetRowData /////////////////////////////
+
+template <DLDeviceType XPU, typename IdType, typename DType>
+NDArray CSRGetRowData(CSRMatrix csr, int64_t row) {
+  CHECK(CSRHasData(csr)) << "missing data array";
+  CHECK(row >= 0 && row < csr.num_rows) << "Invalid row index: " << row;
+  const int64_t len = impl::CSRGetRowNNZ<XPU, IdType>(csr, row);
+  const IdType* indptr_data = static_cast<IdType*>(csr.indptr->data);
+  const int64_t offset = indptr_data[row] * sizeof(DType);
+  return csr.data.CreateView({len}, csr.data->dtype, offset);
+}
+
+template NDArray CSRGetRowData<kDLCPU, int32_t, int32_t>(CSRMatrix, int64_t);
+template NDArray CSRGetRowData<kDLCPU, int64_t, int64_t>(CSRMatrix, int64_t);
+
+///////////////////////////// CSRGetData /////////////////////////////
+
+template <DLDeviceType XPU, typename IdType, typename DType>
+NDArray CSRGetData(CSRMatrix csr, int64_t row, int64_t col) {
+  CHECK(CSRHasData(csr)) << "missing data array";
+  // TODO(minjie): use more efficient binary search when the column indices
```

Comment:
This is probably redundant.

Reply:
The comment sentence is broken. Will fix that. Ideally, if the column indices are sorted for each neighborhood, a binary search could be used instead of the current linear search.

Reply:
Done.